builtin/providers/aws/auth_helpers.go

```diff
@@ -0,0 +1,170 @@
+	}
+
+	log.Printf("[DEBUG] Getting account ID via iam:GetUser failed: %s", err)
+	log.Println("[DEBUG] Trying to get account ID via iam:ListRoles instead")
+	outRoles, err := iamconn.ListRoles(&iam.ListRolesInput{
+		MaxItems: aws.Int64(int64(1)),
+	})
+	if err != nil {
+		return "", fmt.Errorf("Failed getting account ID via 'iam:ListRoles': %s", err)
+	}
+
+	if len(outRoles.Roles) < 1 {
+		return "", fmt.Errorf("Failed getting account ID via 'iam:ListRoles': No roles available")
+	}
+
+	return parseAccountIdFromArn(*outRoles.Roles[0].Arn)
+}
+
+func parseAccountIdFromArn(arn string) (string, error) {
+	parts := strings.Split(arn, ":")
+	if len(parts) < 5 {
+		return "", fmt.Errorf("Unable to parse ID from invalid ARN: %q", arn)
+	}
+	return parts[4], nil
+}
+
+// AWS will add a "Server: EC2" header value to the metadata response.
+// This allows checking the headers for this value to ensure
+// something else didn't just happen to be listening on that IP:Port
+func responseContainsEC2Header(resp *http.Response) bool {
+	return resp.Header["Server"] != nil && strings.Contains(resp.Header["Server"][0], "EC2")
+}
+
+// This function is responsible for reading credentials from the
+// environment in the case that they're not explicitly specified
+// in the Terraform configuration.
+func getCreds(key, secret, token, profile, credsfile string) *awsCredentials.Credentials {
+	// build a chain provider, lazy-evaulated by aws-sdk
+	providers := []awsCredentials.Provider{
+		&awsCredentials.StaticProvider{Value: awsCredentials.Value{
+			AccessKeyID:     key,
+			SecretAccessKey: secret,
+			SessionToken:    token,
+		}},
+		&awsCredentials.EnvProvider{},
+		&awsCredentials.SharedCredentialsProvider{
+			Filename: credsfile,
+			Profile:  profile,
+		},
+	}
+
+	// We only look in the EC2 metadata API if we can connect
+	// to the metadata service within a reasonable amount of time
+	metadataURL := buildMetadataUrl("/latest")
+	c := http.Client{
+		Timeout: 100 * time.Millisecond,
+	}
+
+	r, err := c.Get(metadataURL)
+	// Flag to determine if we should add the EC2Meta data provider. Default false
+	var useIAM bool
+	if err == nil {
+		if responseContainsEC2Header(r) {
+			useIAM = true
+		}
+	}
+
+	if useIAM {
+		log.Printf("[DEBUG] EC2 Metadata service found, adding EC2 Role Credential Provider")
+		providers = append(providers, &ec2rolecreds.EC2RoleProvider{
+			Client: ec2metadata.New(session.New(&aws.Config{
+				Endpoint: aws.String(metadataURL),
+			})),
+		})
+	} else {
+		log.Printf("[DEBUG] EC2 Metadata service not found, not adding EC2 Role Credential Provider")
+	}
+	return awsCredentials.NewChainCredentials(providers)
+}
+
+func buildMetadataUrl(path string) string {
+	url := os.Getenv("AWS_METADATA_URL")
+	if url != "" {
+		log.Printf("[WARN] Using deprecated ENV variable AWS_METADATA_URL, " +
+			"use AWS_METADATA_HOST instead to set the hostname only")
+		return url
```

Comment:
> `AWS_METADATA_URL` used to be `ts.URL+"/latest"`

Correct, that is why am I treating it differently and just take the whole URL, instead of building it.
That function is used in two places, have a look into the full diff - we call this API first when deciding whether to use EC2Role provider at all and then I used it here (with different path) to get the role ARN and Account ID from it.